### ISSUE TITLE
Proposal: _.while

### ISSUE THREAD

**Issue by Nephylhim** [lodash/lodash#4262]
Hi!

TL;DR:
```javascript
_.while(collection, predicate, func)
```
`func` is called as long as predicate returns `true`

___________________
Use case:

I'm currently working with nested objects:
```javascript
const someOtherArgs = { text: 'hey Lodash!', value: 42 }

const elements = [
  { type: 'something', ...someOtherArgs },
  {
    type: 'wrapper',
    children: [
      { type: 'somethingElse', ...someOtherArgs },
      {
        type: 'anotherWrapper',
        children: [
          { type: 'anItem' },
          { type: 'anotherItem' }
        ]
      }
    ]
  },
  { type: 'finally', ...someOtherArgs },
]
```

I need to have all my elements at the root level (I want to flatten the children of wrappers, recursively):
```javascript
const output = [
  { type: 'something', ...someOtherArgs },
  { type: 'somethingElse', ...someOtherArgs },
  { type: 'anItem' },
  { type: 'anotherItem' },
]
```

One ~~ugly~~ obvious method came to me:
```javascript
const output = _
    .chain(elements)
    .map(e => { return e.children || e })
    .flatMap()
    .map(e => { return e.children || e })
    .flatMap()
    // .filter({ type: 'somethingElse' }) // further computing...
    .value()
```
"Arrg! Code duplication!" (Yes, sorry for your eyes).
This duplication needs to be called as many times as there are nested levels.
I'm unwilling to let these lines stay in my code for more than a few seconds.
As I haven't found any functions (nor proposal) that would solve my problem, I imagined something like this:
```javascript
const out = _
  .chain(elements)
  .while(
    _.some(['children', _.isArray]), // predicate
    _.flow([ // function
      c => _.map(c, e => { return e.children || e }),
      _.flatMap
    ])
  )
  .value();
```
I know this solution isn't really a lightweight one.
It might not be the appropriate solution and I'm open to suggestions =)

Thanks for your time!

__________________________________
Note (my current workaround):

I use a recursive function added as a mixin. This method avoids code duplication:
(function name isn't contractual)
```javascript
function flattenElements(elements = [], flattenBy = []) {
  return _
    .chain(elements)
    .map(e => {
      for (const index in flattenBy) {
        const key = flattenBy[index];
        if (e[key]) return flattenElements(e[key], flattenBy);
      }

      return e;
    })
    .flatMap()
    .value();
}

_.mixin({
  'flattenElements': flattenElements,
});

const output = _
  .chain(elements)
  .flattenElements(['children'])
  // .filter({ type: 'somethingElse' }) // further computing
  .value();
```

**Comment by Nephylhim** [comment]
Hi @YuriGor!

That seems to perfectly do the job! Thanks!
Though I suppose that `while` could solve other use cases?

(by the way, your code pen wasn't saved)

**Comment by YuriGor** [comment]
Saved it again.

Is your 'while ' supposed to get the result returned by the 'func' and pass it as an argument to the 'func' on next iteration?
This behavior reminds me of generators. 

```js
function* genFlattenChildren(c){
  while(_.some(c,(e)=>_.isArray(e.children))){
    c = _(c).map(e => { return e.children || e })
    .flatMap().value();
    yield c;
  }
}

let gen = genFlattenChildren(elements);
for(let c of gen){
   console.log(c);
}
```
https://codepen.io/yurigor/pen/xegLvM?editors=0010

May be we should think of method which accept on generator instead of two callbacks?
```js
let latestYielded = _.while(c, gen);
```

**Comment by Nephylhim** [comment]
Yes, It should get the result returned by the func. You need it to test your condition.

The generator is interesting. Though `_.while` doesn't seem an appropriate name anymore.
I think these methods could allow expressing complex algorithms a lot more easily.

**Comment by YuriGor** [comment]
May be name _.forOf will be good, same as in native js, if it's supposed to accept iterators\generators.

Internally it can rely on native for..of of it's available, otherwise use it's own implementation.

**Comment by falsyvalues** [comment]
Hi @Nephylhim!
As usual we'll keep an eye on [the popularity of the request](https://github.com/lodash/lodash/blob/master/.github/CONTRIBUTING.md#feature-requests).
